### ISSUE TITLE
KeyedService.AnyKey is not usable with ServiceKeyAttribute on `object`-typed arguments due to type validation

### ISSUE THREAD

**Issue by DillonN** [dotnet/runtime#93438]
### Description

I'm trying to use the new Keyed services feature, but it looks like the `KeyedService.AnyKey` piece is not functional when used in conjunction with `ServiceKeyAttribute`?

E: Issue only occurs when trying to use a base class for the `ServiceKeyAttribute` argument, e.g. in the case below where `object` is the argument.

Whenever I try, all combination of types leads to an exception:

```
System.InvalidOperationException: 'The type of the key used for lookup doesn't match the type in the constructor parameter with the ServiceKey attribute.'
```

This appears to be due to a check that requires the parameter's type to be the same as the one on the registration:

https://github.com/dotnet/runtime/blob/0b25e38ad32a69cd83ae246104b32449203cc71c/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/CallSiteFactory.cs#L580-L583

However, this is impossible because `KeyedService.AnyKey` uses a dedicated private type and is only exposed as `object`.

I feel like I'm missing something here, but how does `KeyedService.AnyKey` work with that strict type check in place?

### Reproduction Steps

Create a new console app under .NET 8 RC2, add the Microsoft.Extensions.DependencyInjection package, and paste this into `Program.cs`:

```csharp
using Microsoft.Extensions.DependencyInjection;

var services = new ServiceCollection();
services.AddKeyedTransient<Service>(KeyedService.AnyKey);

// This line throws
// System.InvalidOperationException: 'The type of the key used for lookup doesn't match the type in the constructor parameter with the ServiceKey attribute.'
using var provider = services.BuildServiceProvider(new ServiceProviderOptions { ValidateOnBuild = true });

// If ValidateOnBuild is false, this line also throws
// System.InvalidOperationException: 'The type of the key used for lookup doesn't match the type in the constructor parameter with the ServiceKey attribute.'
var service = provider.GetKeyedService<Service>("test");

public class Service
{
    public Service([ServiceKey] object serviceKey)
    { }
}
```

### Expected behavior

* When using `KeyedService.AnyKey` for a service that uses the `[ServiceKey]` attribute, building the service provider should succeed without validation failures
* When using `KeyedService.AnyKey` for a service that uses the `[ServiceKey]` attribute, resolving the service using any key input should return the registered implementation and fill in the service key value for the marked parameter

### Actual behavior

In both of the above points, a `System.InvalidOperationException` is thrown instead

### Known Workarounds

I can workaround by using a factory service registration instead:

```csharp
using Microsoft.Extensions.DependencyInjection;

var services = new ServiceCollection();
services.AddKeyedTransient<Service>(KeyedService.AnyKey, (_, key) => new Service(key));

// Validation succeeds now
using var provider = services.BuildServiceProvider(new ServiceProviderOptions { ValidateOnBuild = true });

// Resolving service also succeeds now
var service = provider.GetKeyedService<Service>("test");

public class Service
{
    public Service(object serviceKey)
    { }
}
```

### Configuration

```
.NET SDK:
 Version:   8.0.100-rc.2.23502.2
 Commit:    0abacfc2b6

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.23550
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100-rc.2.23502.2\

.NET workloads installed:
<removed for brevity>

Host:
  Version:      8.0.0-rc.2.23479.6
  Architecture: x64
  Commit:       0b25e38ad3

.NET SDKs installed:
  8.0.100-rc.2.23502.2 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
<removed for brevity>
  Microsoft.NETCore.App 8.0.0-rc.2.23479.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
<removed for brevity>

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

**Comment by steveharter** [comment]
PTAL @benjaminpetit

**Comment by benjaminpetit** [comment]
`provider.GetKeyedService<Service>("test");` will not throw if you change the type of the key in your constructor to `string` instead of `object`. The type check might be a bit too restrictive here...

For `services.BuildServiceProvider(new ServiceProviderOptions { ValidateOnBuild = true });`, it seems to be a bug yes, I will investigate.

**Comment by DillonN** [comment]
@benjaminpetit thanks, I didn't notice that, was too focused with having `object` for the key type. I guess with the current check, doing that does not make sense? As the only thing I can get to work is `provider.GetKeyedService<Service>(new object());`. Any type that is not exactly `System.Object` will fail.

**Comment by benjaminpetit** [comment]
Yes, the type check is strict: if you are asking for an `object` in your constructor, the key provided must be an `object`. If you are asking for a `string`, then only a `string` can be provided. But we could allow inheritance IMO (your example could work for example).

**Comment by steveharter** [comment]
@benjaminpetit based on
>provider.GetKeyedService<Service>("test"); will not throw if you change the type of the key in your constructor to string instead of object. The type check might be a bit too restrictive here...

> For services.BuildServiceProvider(new ServiceProviderOptions { ValidateOnBuild = true });, it seems to be a bug yes, I will investigate.

and then later

> Yes, the type check is strict: if you are asking for an object in your constructor, the key provided must be an object. If you are asking for a string, then only a string can be provided. But we could allow inheritance IMO (your example could work for example).

I assume the type check is strict by design, but we could loosen it up by supporting inheritance?

**Comment by drolevar** [comment]
@steveharter Just bumped into the issue myself. I guess the type check is not necessary when AnyKey is used. Because semantically AnyKey is exactly what it is - any key.

Another (stricter) approach could be limit AnyKey usage to object-typed parameters and to to introduce a generic AnyKeyObj for other parameters. But that would be a breaking change.

**Comment by drpdrp** [comment]
I'm experiencing inverted problem with this. Trying to get a list of registered keyed services but getting na empty list.

registration:
`services.AddKeyedSingleton<IMyType, MyType1>(typeof(MyType1).ToString());`

`var list = serviceProvider.GetKeyedServices<IMyType>(KeyedService.AnyKey)`

**Comment by drpdrp** [comment]
I'm experiencing inverted problem. Trying to get a all registered keyed services, but getting empty list. How can I get all registered keyed services from service provider?

registration:
`services.AddKeyedSingleton<IMyType, MyType1>(typeof(MyType1).ToString());`
`services.AddKeyedSingleton<IMyType, MyType2>(typeof(MyType2).ToString());`

access:
`var list = provider.GetKeyedServices<IMyType>(KeyedService.AnyKey) // getting []`

this works:
`provider.GetRequiredKeyedService<IMyType>(typeof(MyType1).ToString())`
`provider.GetRequiredKeyedService<IMyType>(typeof(MyType2).ToString())`

**Comment by drpdrp** [comment]
I'm experiencing inverted problem. Trying to get all registered keyed services, but getting empty list. How can I get all registered keyed services from service provider?

registration:
`services.AddKeyedSingleton<IMyType, MyType1>(typeof(MyType1).ToString());`
`services.AddKeyedSingleton<IMyType, MyType2>(typeof(MyType2).ToString());`

access:
`var list = provider.GetKeyedServices<IMyType>(KeyedService.AnyKey) // getting []`

this works:
`provider.GetRequiredKeyedService<IMyType>(typeof(MyType1).ToString())`
`provider.GetRequiredKeyedService<IMyType>(typeof(MyType2).ToString())`

**Comment by davidfowl** [comment]
This is something we should fix in .NET 10.

**Comment by DillonN** [comment]
It would be a welcome change! I still have the factory registration workaround setup in a few places

**Comment by steveharter** [comment]
Based on the discussions above, supporting inheritance for a service constructor parameter using `ServiceKeyAttribute` seems like the correct approach. If that is not feasible for some reason, a fallback approach would be to just support `object`.

**Comment by steveharter** [comment]
@davidfowl 

> This is something we should fix in .NET 10.

Do you mean the comment just above yours regarding `GetKeyedServices<IMyType>(KeyedService.AnyKey)` (which was fixed in v9) or the original issue here (support for `object` as parameter type if `AnyKey` is used)?

**Comment by georg138** [comment]
The check also fails when using `string` as key type and only using `string` as keys with `FromKeyedServices`. However, the check fails when comparing the type of the constructor parameter (`string`) to the type of `KeyedService.AnyKey`, which resolves to `object`.
